<?xml version="1.0" encoding="utf-8"?>
<!--
Launch file for Selective ICP with rviz
-->

<launch>

  <!-- Proper Output Info -->
  <env name="ROSCONSOLE_FORMAT" value="[${severity}] (${node}) [${time}]: ${message}\n" />

  <param name="/use_sim_time" value="true" />
    <arg name="cad_file" default="$(find cpt_selective_icp)/resources/cla.json" />
    <arg name="rosbag" />
    <arg name="logprefix" />


    <!--node name="mesh_positioning"
          pkg="cpt_selective_icp"
          type="position_cad_with_marker.py"
          output="screen">
        <param name="marker_parent_frame" value="map" />
        <param name="marker_frame_name" value="marker" />
      </node-->


    <node pkg="background_foreground_segmentation"
          type="pose_logger.py"
          name="poselogger">
        <param name="filename" value="/home/blumh/asl/rss_2021_data/$(arg logprefix)_icp.csv" />
    </node>
    <node pkg="background_foreground_segmentation"
          type="point_logger.py"
          name="leicalogger">
        <param name="filename" value="/home/blumh/asl/rss_2021_data/$(arg logprefix)_leica.csv" />
        <param name="overwrite" value="false" />
    </node>

    <!-- Start Dataset Creator node for each camera -->
    <node pkg="background_foreground_segmentation" type="background_foreground_segmentation_node"
          name="dataset_creator_cam0" output="screen">
        <rosparam command="load" file="$(find background_foreground_segmentation)/config/cla_dataset_creator_in_memory.yaml"/>
        <param name="outputFolder" value="/tmp/"/>
        <param name="cameraImageTopic" value="/pickelhaube_resized/cam0/image"/>
        <param name="cameraInfoTopic" value="/pickelhaube_resized/cam0/camera_info"/>
        <param name="cameraFrame" value="/cam0_rectified_frame"/>
        <param name="outputTopic" value="/meshdist/cam0/original"/>
        <param name="labelsTopic" value="/meshdist/cam0/projected_pc"/>
        <param name="distanceTopic" value="/meshdist/cam0/projected_pc_distance"/>
    </node>

    <node pkg="background_foreground_segmentation" type="background_foreground_segmentation_node"
          name="dataset_creator_cam1" output="screen">
        <rosparam command="load" file="$(find background_foreground_segmentation)/config/cla_dataset_creator_in_memory.yaml"/>
        <param name="outputFolder" value="/tmp/"/>
        <param name="cameraImageTopic" value="/pickelhaube_resized/cam1/image"/>
        <param name="cameraInfoTopic" value="/pickelhaube_resized/cam1/camera_info"/>
        <param name="cameraFrame" value="/cam1_rectified_frame"/>
        <param name="outputTopic" value="/meshdist/cam1/original"/>
        <param name="labelsTopic" value="/meshdist/cam1/projected_pc"/>
        <param name="distanceTopic" value="/meshdist/cam1/projected_pc_distance"/>
    </node>

    <node pkg="background_foreground_segmentation" type="background_foreground_segmentation_node"
          name="dataset_creator_cam2" output="screen">
        <rosparam command="load" file="$(find background_foreground_segmentation)/config/cla_dataset_creator_in_memory.yaml"/>
        <param name="outputFolder" value="/tmp/"/>
        <param name="cameraImageTopic" value="/pickelhaube_resized/cam2/image"/>
        <param name="cameraInfoTopic" value="/pickelhaube_resized/cam2/camera_info"/>
        <param name="cameraFrame" value="/cam2_rectified_frame"/>
        <param name="outputTopic" value="/meshdist/cam2/original"/>
        <param name="labelsTopic" value="/meshdist/cam2/projected_pc"/>
        <param name="distanceTopic" value="/meshdist/cam2/projected_pc_distance"/>
    </node>

    <node pkg="background_foreground_segmentation"
          type="label_aggregator.py"
          name="label_writer"
          output="screen">
        <rosparam>
        publish_labels: false
        store_labels: true
        label_frequency: 3
        image_topics:
          - - /pickelhaube_resized/cam0/image
            - /meshdist/cam0/projected_pc
            - /meshdist/cam0/projected_pc_distance
          - - /pickelhaube_resized/cam1/image
            - /meshdist/cam1/projected_pc
            - /meshdist/cam1/projected_pc_distance
          - - /pickelhaube_resized/cam2/image
            - /meshdist/cam2/projected_pc
            - /meshdist/cam2/projected_pc_distance
        out_topics:
          - /meshdist/cam0/labels
          - /meshdist/cam1/labels
          - /meshdist/cam2/labels
        label_options:
          downsamplingFactor: 1
          useSuperpixel: true
          agreement: false
          foregroundTrustRegion: true
          fgBgThreshold: 100
          numberOfSuperPixel: 1000
          stdDevThreshold: 1
        </rosparam>
        <param name="label_path" value="/home/blumh/asl/rss_2021_data/$(arg logprefix)/" />
    </node>

    <node type="rviz" name="rviz" pkg="rviz" args="-d $(find cpt_selective_icp)/launch/rviz.rviz" />

    <!-- should be started last for handover to ICP-->
    <include file="$(find background_foreground_segmentation)/launch/pickelhaube/pickelhaube_versavis_from_bag.launch">
        <arg name="rosbag" value="$(arg rosbag)" />
        <arg name="rate" value="0.1" />
    </include>

    <node name="mapper"
          type="cpt_selective_icp_node"
          pkg="cpt_selective_icp"
          output="screen" >
          <!--launch-prefix="tmux split-window" -->
    <rosparam command="load" file="$(find background_foreground_segmentation)/launch/paper_experiments/mapper_parameters.yaml" />
                <param name="standaloneICP" value="true" />
                <param name="scanTopic" value="/rslidar_points" />
        <param name="cadTopic" value="mesh_publisher/mesh_out" />
        <param name="icpConfig"
               value="$(find background_foreground_segmentation)/launch/paper_experiments/office_icp.yaml" />
        <param name="selectiveIcpConfig"
               value="$(find background_foreground_segmentation)/launch/paper_experiments/selective_icp.yaml" />
        <param name="inputFiltersConfig"
               value="$(find background_foreground_segmentation)/launch/paper_experiments/office_input_filters.yaml" />
        <param name="mapPostFiltersConfig"
               value="$(find background_foreground_segmentation)/launch/paper_experiments/map_post_filter.yaml" />
        <param name="mapPreFiltersConfig"
               value="$(find background_foreground_segmentation)/launch/paper_experiments/map_pre_filter.yaml" />
        <param name="path"
                 value="$(find cpt_selective_icp)" />
    </node>
    <node name="mesh_publisher"
          pkg="cpt_utils"
          type="mesh_publisher_node"
          output="screen"
          launch-prefix="bash -c 'sleep 4; $0 $@'">
        <param name="publish_on_start" value="true"/>
        <param name="default_filename" value="$(arg cad_file)"/>
        <param name="frame_name" value="marker"/>
      </node>
</launch>
